core/src/main/java/org/jdbi/v3/core/mapper/reflect/ImmutableMapper.java

```diff
@@ -194,41 +166,44 @@ public T map(final ResultSet rs, final StatementContext ctx) throws SQLException
+            constructorTemp.setAccessible(true);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+        }
+        final Constructor<MethodHandles.Lookup> constructor = constructorTemp;
+
         return (r, c) -> {
-            Object builder = null;
-            try {
-                builder = builderCreationMethod.invoke(null);
-            } catch (IllegalAccessException | InvocationTargetException e) {
-                throw new MappingException("Could not create Builder", e);
-            }
+            final Map<Method, Object> invocations = new HashMap<>();
             for (int i = 0; i < mappers.size(); i++) {
                 RowMapper<?> mapper = mappers.get(i);
                 PropertyDescriptor property = properties.get(i);
-
                 Object value = mapper.map(r, ctx);
-
-                try {
-                    MethodHandles.lookup().unreflect(property.getWriteMethod()).bindTo(builder).invokeWithArguments(value);
-                } catch (IllegalAccessException e) {
-                    throw new MappingException("Setter for the builder should be public", e);
-                } catch (RuntimeException e) {
-                    throw e;
-                } catch (Throwable e) {
-                    throw new MappingException("Immutable build failed with exception", e);
-                }
+                invocations.put(property.getReadMethod(), value);
             }
-            try {
-                return (T) MethodHandles.lookup().findVirtual(builder.getClass(),
-                    "build",
-                    MethodType.methodType(implementation)).bindTo(builder).invokeWithArguments();
-            } catch (NoSuchMethodException | IllegalAccessException e) {
-                throw new MappingException("Expected to find a Builder with public build() method", e);
-            } catch (Throwable e) {
-                if (e instanceof RuntimeException) {
-                    throw (RuntimeException) e;
+
+            return (T) Proxy.newProxyInstance(this.getClass().getClassLoader(), new Class[]{type}, new InvocationHandler() {
+                private final Map<Method, Object> invocationMap = invocations;
+
+                @Override
+                public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                    Object returnObj = invocationMap.get(method);
+                    if (returnObj == null) {
+                        if (method.isDefault() && constructor != null) {
+                            final Class<?> declaringClass = method.getDeclaringClass();
+                            return constructor.newInstance(declaringClass, MethodHandles.Lookup.PRIVATE)
+                                .unreflectSpecial(method, declaringClass)
+                                .bindTo(proxy)
+                                .invokeWithArguments(args);
```

Comment:
Are you running on Java 9? I think you'll run into problems with `unreflectSpecial` on default methods in Java 8.